homeassistant/components/todo/intent.py

```diff
@@ -0,0 +1,69 @@
+"""Intents for the todo integration."""
+from __future__ import annotations
+
+import voluptuous as vol
+
+from homeassistant.components.shopping_list import EVENT_SHOPPING_LIST_UPDATED
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import intent
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_component import EntityComponent
+
+from . import DOMAIN, TodoItem, TodoListEntity
+
+INTENT_ADD_ITEM = "HassShoppingListAddItem"
+
+
+async def async_setup_intents(hass: HomeAssistant) -> None:
+    """Set up the todo intents."""
+    intent.async_register(hass, AddItemIntent())
+
+
+class AddItemIntent(intent.IntentHandler):
+    """Handle AddItem intents."""
+
+    intent_type = INTENT_ADD_ITEM
+    slot_schema = {"item": cv.string, vol.Optional("list"): cv.string}
+
+    async def async_handle(self, intent_obj: intent.Intent):
+        """Handle the intent."""
+        hass = intent_obj.hass
+
+        slots = self.async_validate_slots(intent_obj.slots)
+        item = slots["item"]["value"]
+
+        component: EntityComponent[TodoListEntity] = hass.data[DOMAIN]
+        list_entities: list[TodoListEntity] = list(component.entities)
+        if not list_entities:
+            raise intent.IntentHandleError("No list entities")
+
+        target_list: TodoListEntity | None = None
+
+        if "list" in slots:
+            # Add to a list by name
+            list_name = slots["list"]["value"]
+
+            # Find matching list
+            matching_states = intent.async_match_states(
+                hass, name=list_name, domains=[DOMAIN]
+            )
+            for list_state in matching_states:
+                target_list = component.get_entity(list_state.entity_id)
+                if target_list is not None:
+                    break
+
+            if target_list is None:
+                raise intent.IntentHandleError(f"No list named {list_name}")
+        else:
+            # Add to the first list
```

Comment:
I've been thinking about this. For backwards compat today we could just target the list that is created by the shopping list integration.

Reply:
What if we add an optional `domain_hint` slot to the intent that defaults to `shopping_list`? This causes the todo entity search to prefer a specific domain, both with and without a name.

This would let us support sentences like "add X to todoist" (with `domain_hint` set to `todoist` by the sentence data) and it would add X to the first todoist list. It also provides a way to target lists with the same name from different domains.

Reply:
I guess that results in the same backwards compat behavior, so I guess ok? I don't see how the domain hint would get set in other situations (as one cannot say "todoist" and expect STT to pick that up)